### ISSUE TITLE
No error is given when a caller incorrectly attempts to unpack a tuple from a function that doesn't return a tuple

### ISSUE THREAD

**Issue by bricker** [microsoft/pyright#6108]
Given a function that returns a non-Tuple type, attempting to unpack the function's return value as if it were a tuple (i.e. multiple assignment) is a runtime error. Pyright doesn't warn the user about this.

**Example**

The following code results in a runtime error, but isn't caught by Pyright:

```python
def foo() -> str:
    return "foo"

bar, baz = foo()  # ValueError: too many values to unpack (expected 2)
```

I expect Pyright to warn me that the function `foo` returns a string, so multiple assignment isn't possible.

**Versions**

- pyright CLI == 1.1.330
- Pylance VSCode extension == v2023.10.10
- Python VSCode extension == v2023.18.0

**Comment by erictraut** [comment]
This is by design.

Unpacking isn't limited to tuple types. The unpack operator works with any iterable type. A `str` is an iterable type, so it can be unpacked. The length of an iterable type can't generally be known statically. Type checkers like pyright and mypy err on the side of not generating an error in such cases (at the risk of it being a false negative) rather than generating what is likely a false positive much of the time.

**Comment by bricker** [comment]
I see. If the function in my example returned a string of length two, then no runtime error occurs. My mistake, and thank you for the explanation!
